zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,276 @@
+    String[] children = childrenSet.toArray(new String[0]);
+
+    // Find child nodes that's in zk server but not in restored data tree
+    try {
+      List<String> destChildren = zk.getChildren(path, false);
+      for (String destChild : destChildren) {
+        if (!childrenSet.contains(destChild)) {
+          String nodePath = path + "/" + destChild;
+          messages.add(nodePath + ": This node does not exist in the restored data tree.");
+        }
+      }
+    } catch (Exception e) {
+      skipNodeOrStopRestoration(path, e);
+    }
+
+    // Traverse the child nodes of this node
+    for (String child : children) {
+      traverseNode(zk, dataTree, path + "/" + child, false);
+    }
+  }
+
+  /**
+   * Restore the znode in the zk server, currently only doing creation of non-existing nodes,
+   * and only supports persistent nodes
+   * @param zk A client connection to zk server
+   * @param dataTree The restored zk data tree
+   * @param path The znode path to restore
+   * @param targetNode True if the node being restored is the target znode that user specified;
+   *                   false if it's a child node of the target node
+   * @return True if the node is successfully restored, false if the node is skipped
+   */
+  private boolean restoreNode(ZooKeeper zk, DataTree dataTree, String path, boolean targetNode) {
+    DataNode node = dataTree.getNode(path);
+    try {
+      if (zk.exists(path, false) == null) {
+        Stat stat = new Stat();
+        node.copyStat(stat);
+        if (stat.getEphemeralOwner() != 0) {
+          messages.add(path
+              + ": The node is an ephemeral node, which is not supported in spot restoration.");
+        } else {
+          return createNode(path, dataTree);
+        }
+      } else if (targetNode) {
+        // If it is the target node that the user wants to restore,
+        //we don't skip it just because it already exists in the server,
+        //because we already get user's confirmation to restore this node
+        //and we need to get to its child nodes
+        zk.setData(path, node.getData(), -1);
+        return true;
+      } else {
+        messages.add(path + ": The node already exists in zk server.");
+      }
+    } catch (Exception e) {
+      skipNodeOrStopRestoration(path, e);
+    }
+    return false;
+  }
+
+  private void print(String message) {
+    System.out.println(message);
+  }
+
+  @VisibleForTesting
+  protected boolean getUserConfirmation(String requestMsg, String yesMsg, String noMsg) {
+    Scanner scanner = new Scanner(System.in);
+    while (true) {
+      print(requestMsg);
+      String input = scanner.nextLine().toLowerCase();
+      switch (input) {
+        case "yes":
+          print(yesMsg);
+          return true;
+        case "no":
+          print(noMsg);
+          return false;
+        default:
+          print("Could not recognize the input: " + input + ". Please try again.");
+          break;
```

Comment:
NIT: Shall we make a return here as well? Or we can try 3 times and fail it? Because if this tool is leveraged by some machine tool with continuous unexpected input. The process will be hanging there and keep retrying .